integrations/kujira/packages/abstract-adapter/src/dex.rs

```diff
@@ -321,3 +321,238 @@ pub fn decimal2decimal256(dec_value: Decimal) -> StdResult<Decimal256> {
+                    Asset::new(AssetInfo::native(DEMO), amount_demo),
+                    Asset::new(AssetInfo::native(KUJI), amount_kuji),
+                ],
+                Some(max_spread()),
+            )
+            .unwrap();
+
+        expect_eq(
+            vec![wasm_execute(
+                POOL_CONTRACT,
+                &bow::market_maker::ExecuteMsg::Deposit {
+                    max_slippage: Some(max_spread()),
+                    callback: None,
+                },
+                vec![coin(amount_demo, DEMO), coin(amount_kuji, KUJI)],
+            )
+            .unwrap()
+            .into()],
+            msgs,
+        )
+        .unwrap();
+    }
+
+    #[test]
+    fn provide_liquidity_one_side() {
+        let amount_demo = 100_000u128;
+        let amount_kuji = 0u128;
+        let msgs = create_setup()
+            .test_provide_liquidity(
+                pool_addr(),
+                vec![
+                    Asset::new(AssetInfo::native(DEMO), amount_demo),
+                    Asset::new(AssetInfo::native(KUJI), amount_kuji),
+                ],
+                Some(max_spread()),
+            )
+            .unwrap();
+
+        // There should be a swap before providing liquidity
+        // We can't really test much further, because this unit test is querying mainnet liquidity pools
+        expect_eq(
+            wasm_execute(
+                SWAP_CONTRACT,
+                &fin::ExecuteMsg::Swap {
+                    offer_asset: Some(coin(amount_demo / 2u128, DEMO)),
+                    belief_price: None,
+                    max_spread: Some(decimal2decimal256(max_spread()).unwrap()),
+                    to: None,
+                    callback: None,
+                },
+                coins(amount_demo / 2u128, DEMO),
+            )
+            .unwrap()
+            .into(),
+            msgs[0].clone(),
+        )
+        .unwrap();
+    }
+
+    #[test]
+    fn provide_liquidity_symmetric() {
+        let amount_demo = 100_000u128;
+        let msgs = create_setup()
+            .test_provide_liquidity_symmetric(
+                pool_addr(),
+                Asset::new(AssetInfo::native(DEMO), amount_demo),
+                vec![AssetInfo::native(KUJI)],
+            )
+            .unwrap();
+
+        assert_eq!(msgs.len(), 1);
+        assert_eq!(get_wasm_addr(msgs[0].clone()), POOL_CONTRACT);
+
+        let unwrapped_msg: bow::market_maker::ExecuteMsg = get_wasm_msg(msgs[0].clone());
+        match unwrapped_msg {
+            bow::market_maker::ExecuteMsg::Deposit {
+                max_slippage,
+                callback,
+            } => {
+                assert_eq!(max_slippage, None);
+                assert_eq!(callback, None);
+            }
+            _ => panic!("Expected a provide liquidity variant"),
+        }
+
+        let funds = get_wasm_funds(msgs[0].clone());
+        assert_eq!(funds.len(), 2);
+        assert_eq!(funds[0], coin(amount_demo, DEMO),);
```

Comment:
Weird that there's a trailing comma here and the linter didn't yell at you.

Reply:
This sorting stuff is craaaazy, why didn't we know that ? That's weird